jbi/services.py

```diff
@@ -127,6 +129,68 @@ def _all_jira_projects_visible(jira, actions: Actions) -> bool:
     return not missing_projects
 
 
+def _all_jira_projects_permissions(jira, actions: Actions):
+    """Fetches and validates that required permissions exist for the configured projects"""
+    all_projects_perms = _fetch_jira_project_permissions(actions, jira)
+    return _validate_jira_permissions(all_projects_perms)
+
+
+def _fetch_jira_project_permissions(actions, jira):
+    """Fetches permissions for the configured projects"""
+    required_perms_by_project = {
+        action.parameters["jira_project_key"]: action.required_jira_permissions
+        for action in actions
+        if "jira_project_key" in action.parameters
+    }
+
+    all_projects_perms = {}
+    # Query permissions for all configured projects in parallel threads.
+    with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
+        futures_to_projects = {
+            executor.submit(
+                jira.get_permissions,
+                project_key=project_key,
+                permissions=",".join(required_permissions),
+            ): project_key
+            for project_key, required_permissions in required_perms_by_project.items()
+        }
+        # Obtain futures' results unordered.
+        for future in concurrent.futures.as_completed(futures_to_projects):
+            project_key = futures_to_projects[future]
+            response = future.result()
+            all_projects_perms[project_key] = (
+                required_perms_by_project[project_key],
+                response["permissions"],
+            )
+    return all_projects_perms
+
+
+def _validate_jira_permissions(all_projects_perms):
+    """Validates permissions for the configured projects"""
+    misconfigured = []
+    for project_key, (required_perms, obtained_perms) in all_projects_perms.items():
+        missing = required_perms - set(obtained_perms.keys())
+        not_given = set(
+            entry["key"]
+            for entry in obtained_perms.values()
+            if not entry["havePermission"]
+        )
```

Comment:
We could simplify this a bit if we had the guarantee that we get an entry for each permission that we specified in the `jira.get_permissions()` call. But I can't find it in the docs, therefore we have this double-checking here (list + `havePermission` key.